run.py

```diff
@@ -0,0 +1,187 @@
+        return '1' if value else ''
+    else:
+        return str(value)
+
+
+def list_filters(path):
+    for filename in glob.glob(path, recursive=True):
+        try:
+            with open(filename) as fh:
+                defaults = {
+                    "name": os.path.splitext(os.path.basename(filename))[0],
+                    "basedir": os.path.dirname(filename)
+                }
+                yield {**defaults, **json.load(fh)}
+        except Exception as e:
+            print(f"Could not parse {filename}: {e}", file=sys.stderr)
+
+
+def babysit_child(child, name, print_queue, ctrl_queue):
+    prefix = f'[{name}] '.encode()
+
+    for line in child.stderr:
+        print_queue.put(prefix + line)
+
+    child.wait()
+
+    print_queue.put(f'[run.py] {name} exited with status code {child.returncode}\n'.encode())
+
+    ctrl_queue.put(child.returncode)
+
+
+def print_lines(queue, fout):
+    while True:
+        line = queue.get()
+        if line is None:
+            break
+        fout.buffer.write(line)
+
+
+def main(argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--filters', '-f', type=str, default='./filters')
+    parser.add_argument('--input', '-i', type=argparse.FileType('rb'))
+    parser.add_argument('--output', '-o', type=argparse.FileType('wb'), default=sys.stdout.buffer)
+    parser.add_argument('--basedir', '-b', type=str, help='Directory to look for data files when -i is not used')
+    parser.add_argument('pipeline', metavar='PIPELINE', type=argparse.FileType('r'))
+    parser.add_argument('languages', metavar='LANG', type=str, nargs=2)
+
+    args = parser.parse_args(argv)
+
+    # default search path for the data files is next to the configuration file
+    # which is the default save location for empty-train.
+    if not args.basedir:
+        args.basedir = os.path.dirname(args.pipeline.name)
+
+    pipeline = json.load(args.pipeline)
+
+    # load all filter definitions (we need to, to get their name)
+    FILTERS = {
+        definition['name']: definition
+        for definition in list_filters(os.path.join(args.filters, '*.json'))
+    }
+
+    # Assert we have all filters we need
+    assert set(step['filter'] for step in pipeline['filters']) - set(FILTERS.keys()) == set()
+
+    children: List[Popen] = []
+
+    babysitters: List[Thread] = []
+
+    print_queue = SimpleQueue() # type: SimpleQueue[Optional[bytes]]
+
+    ctrl_queue = SimpleQueue() # type: SimpleQueue[int]
+
+    # If we're not reading from stdin, read from files and paste them together
+    if not args.input:
+        for filename in pipeline['files']:
+            child = subprocess.Popen(
+                ['gzip', '-cd', filename],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                cwd=args.basedir)
+
+            children.append(child)
+
+            thread = Thread(target=babysit_child, args=[child, f'ungzip {filename}', print_queue, ctrl_queue])
```

Comment:
shouldn't this be `gunzip`

Reply:
It's just for presentation, but I've replaced it to avoid confusion.